[CR]Resistances update (including fire resistance)

Coolthulhu opened the pull request:
Summary:
- Split cut and stab resistance for monsters. Blobs and skeletons are easy to cut, hard to stab.
- Add acid resistance stat to monsters, so that they can be resistant-but-not-immune. For example, skeletons and firemen.
- Add fire resistance to monsters. Firemen will still burn, but not as readily as regular zombies.
- Add fire resistance to player's items. It is based on items' own fire resistance and their environmental protection. Turnout boots will let you run around through small fires.
- Limit "onfire" effect to 10 turns. Before it could stack to some insane durations, despite there being nothing left to burn.
- Refactor armor resistance getters. It is now split (with a `bool` argument) into "armor granted to wearer" and "own resistance to damage". The latter is higher than the former in some cases (items with low environmental protection will resist fire, but not protect from it).
- Add a more "generic" armor getter for `Creature`. The older one was just bashing and cutting and duplicating all the functions for each type after that would get clunky really fast. The new one takes damage type as an argument.
- Simplified acid resistance calculations. The displayed number is the actual damage decrease, so no hidden and confusing armor penetration this time. The resistances are much lower because of that.

[CR] because I might have missed something and because balancing.
It is too easy to get fire resistance. Put on 2 pairs of fire-resistance socks, move them outside your boots and you can stomp around medium fires. Full survivor firegear will let you ignore raging fires.
Any good way of nerfing that? Or maybe it's OK because the firegear-clad player is supposed to be an endgame monstrosity, a Nomex horror with napalm blood?

hitbutton commented:
Balance/Reality wise I think its fine that some protective clothing makes you broadly resistant to direct damage from fire.

The thing that should stop you from just standing around in raging fires (ignoring oxygen deprivation) should be the heat rather than the burning. Various materials can protect you from the latter, but stuff like nomex can only protect you from the former for a short amount of time.

Rivet-the-Zombie commented:
Very nice!

kevingranade commented:
General balance point for fireproof gear is encumbrance and warmth.
Nomex might shield you from heat, but it's trapping the heat you generate
too.

hitbutton commented:
I had a go at testing out my heat limits in this PR.
I wore an entry suit and filled my tile and the tiles surrounding me with 2x4s, then set them on fire and waited.

After 15 minutes my body temperature on the character screen is 911, all body parts scorching, -110% move speed from Hot, -13% from Pain. Base Move 168, Move Speed 25. Full HP in all body parts.

At ~16 minutes I lost control of my character, while frames passed visually for some time (similar to waiting). The program then froze and stopped responding.

I've had a quick peek at the temperature code for something that might cause me to become unable to act but coudn't find anything? It's also possible this behaviour was caused by a lot of tiles completely full of burning 2x4s.

hitbutton commented:
Incidentally - wrapping up material fire resistance and the degree it protects the player from fire seems oversimplified. Items made of steel, kevlar and nomex would all be relatively undamaged by fire, but I'd much rather be wearing nomex in a burning building.

illi-kun commented on src/monstergenerator.cpp:
```diff
@@ -453,6 +455,13 @@ void MonsterGenerator::load_monster(JsonObject &jo)
         newmon->vision_night = jo.get_int("vision_night", 1);
 
+        // Stab armor is optional, use 80% cut armor if it isn't set
+        if( jo.has_int("armor_stab") ) {
```
Isn't it the same to ```newmon->armor_stab = jo.get_int("armor_stab", 0.8f * newmon->armor_cut);```?

chaosvolt commented:
> Add fire resistance to player's items. It is based on items' own fire resistance and their environmental protection. Turnout boots will let you run around through small fires.

Yessss.

hitbutton commented:
In further testing, I popped on an entry suit and went for a stroll through some lava (debug teleporting into an underground rift). I quickly lost all HP from my legs and feet. Immediately afterwards I lost control of my character and the game appeared to lock up. A short time later the game unlocked, and I was informed that I had died, three days later, of dehydration. I still had near full HP in my head and arms, despite three days lying in a pool of lava.

chaosvolt commented:
Convection is overrated.

Coolthulhu commented:
Well, lava is currently handled in a totally different way - it is a trap rather than a field. This means it only "acts" on the turn someone activates it (usually by stepping on it).
It does make "fire attacks" (20 dmg to each foot and leg) against the player, which means fire gear should protect from its damage (getting it to protect from all 20 would be quite hard, though), but the fire damage itself doesn't "paralyze" the player.

The lockup is probably caused by vomiting.
When your head has the "hot" status, you have a chance of vomiting. This chance increases with temperature. Due to math involved, it looks like it can reach 100%:
```
if (one_in(std::min(14500, 15000 - temp_cur[bp_head]))) {
    vomit();
```

That `min` should probably be `max`.

Since vomiting takes 100 moves and your speed was most likely reduced to 25, one vomit (even if nothing comes out) per 4 turns is enough for a lock-up.

On an unrelated note: damn, hallucination code is awful. It has a static bool in it, despite trying to handle NPCs in the same block (using the same static bool at that).
Even worse: a hallucinating NPC would probably summon hallucinations that only the player would see, tormenting the player for using debug commands on them (I don't think NPCs can gain hallucination status in any non-cheaty way).

chaosvolt commented:
Hmm. Given lava also seems to never ignite nearby structures or items, it might be better if if behaved more like fire, or used some hackish method like the gas vent trap to summon flames on its location (possibly best not to allow it to propagate like gas vents, and the field might need to be forced invisible like spacial distortion.)

hitbutton commented:
> That min should probably be max.
A mechanism to avoid the "lockup" would be good, but reducing the incapacitation caused by the extreme head heat would seem to move in the wrong direction though. In reality an internal temp of 50°C (122°F) should basically result in instant death.

I appreciate this merge isn't directly addressing the effects of body temperature, but it's bringing them to light, as it wasn't previously possible to stand around in infernos without dying.

Coolthulhu commented:
> A mechanism to avoid the "lockup" would be good, but reducing the incapacitation caused by the extreme head heat would seem to move in the wrong direction.

Those temperatures don't happen in regular games. You need to get very close to fires to get locked up. This lockup isn't intended.
This should be handled by making extreme heat cause damage, not incapacitation. Incapacitation would make sense if NPCs/minions could rescue the player, until then it would be better to keep player in control and just cause slowdown and damage.

> it wasn't previously possible to stand around in infernos without dying.

The player can get fireproof by getting a proper bionic or mutation. Those have the same limitation: it's fireproofing but not "heatproofing".

chaosvolt commented:
Small issue here: Will this PR do anything about whole "fireproof not heatproof" thing?

Logically a highly-insulating entry suit would essentially trap body heat in but keep external heat out. How would we model that? Have fire resistance set a reduction or cap on how heat from external sources affects the player, maybe? They already have high warmth bonuses, so if they essentially warmed up the body part but mitigated heat buildup from the environment...

Coolthulhu commented:
I wanted to keep it mostly technical because it is already quite big and covers multiple balance changes.
Adding a damage effect to blisters and possibly scorching heat wouldn't be much work, though. It would literally be just json changes (frostbite already deals damage). The only problem is balancing.

chaosvolt commented:
Ah right. True, that might be best to add fleshed out heat changes in a separate PR, yeah. ^^"

kevingranade commented:
There are no such articles that totally isolate you from external
temperature extremes.
Even if there were, you'd cook from your own body heat anyway, the human
body must shed heat to survive.

hitbutton commented:
But a very large point of nomex firefighting gear is that it *does* protect you from radiant heat (for relatively short durations) as it has extremely low thermal conductivity when exposed to high temperatures.

One way of dealing with that would be to treat clothing "warmth" as thermal resistance. Greater resistance would slow the acclimatization of body temperature to environmental temperature, so it would make you warmer in environs up to ~100°F, but slow the process of getting hotter above that temperature.

chaosvolt commented:
Hmm. So unsure how best to handle it in gameplay terms.

Rivet-the-Zombie commented:
>Put on 2 pairs of fire-resistance socks, move them outside your boots and you can stomp around medium fires. Full survivor firegear will let you ignore raging fires.
Any good way of nerfing that? Or maybe it's OK because the firegear-clad player is supposed to be an endgame monstrosity, a Nomex horror with napalm blood?

You'd still cook after a while from the heat, right?

chaosvolt commented:
Yeah. As it is now it would be nice if nomex gear could insulate you somewhat, but you'd realistically still overheat. Unsure of the best way to model that, but that can be figured out in detail another day.

As it is, since this PR doesn't add insulation from heat (far as I can still), it wouldn't be too horridly OP, and fleshing out insulation would allow further balancing later on.

hitbutton commented:
> As it is, since this PR doesn't add insulation from heat (far as I can still), it wouldn't be too horridly OP
I have to disagree. The penalties of heat are excessively lenient right now, and an entry suit + lighter allows you to trivialise the vast majority of enemies in the game without significant risk to yourself.

Coolthulhu commented:
How about giving all monsters without explicit fire resistance a default fire resistance of, say, 50% their cutting armor?

Rivet-the-Zombie commented:
That's good! Alternatively, default fire resistance could be derived from the creature's material type.

Coolthulhu commented:
> Alternatively, default fire resistance could be derived from the creature's material type.

Creatures already do have some effects from material resistance, but in a more ugly, hardcoded way:
In fire field damage section, there is a check for materials. Fleshy critters get +3 damage, veggy (including fungal) get +12, liquid (blobs) get +20, steel get -20.
Steel and liquid creatures can't be set on fire (which deals untyped, not fire damage).

I could move that to default armor calculations too. Or just remove it except for the steel one, since "negative resistances" in flat values are weird as they affect small fires more than raging ones.

Fun fact: flamethrower's initial hit deals piercing, not heat damage.

chaosvolt commented:
> Fun fact: flamethrower's initial hit deals piercing, not heat damage.

Most likely because every ranged weapon in the game does a mundane projectile attack underneath their extra effects.

And yeah, giving monsters a bit more resistance to fire would help with ensuring that the initial boost to fire resistance from an entry suit doesn't make it too easy. Though as it is, you ALREADY can trivialize most zombies with a lighter, it's just that an entry suit means you can afford to use that indoors.

Though you also lose the loot in that building, so there are already good reasons not to abuse that shiny new fire protection. Just one less reason than we have as it is.

Rivet-the-Zombie commented:
>Fun fact: flamethrower's initial hit deals piercing, not heat damage.

>>Most likely because every ranged weapon in the game does a mundane projectile attack underneath their extra effects.

That's the RM451 flamethrower's whole gimmick.

chaosvolt commented:
I thought their gimmick was leaving a line of flame, but no explosion at the target. :V

Rivet-the-Zombie commented:
That too.

hitbutton commented:
> How about giving all monsters without explicit fire resistance a default fire resistance of, say, 50% their cutting armor?
The issue is being able to camp out inside a raging inferno for extended periods of time. IMO Balance should come from negative effects from extremely high temperatures, rather than fireproofing zombies.

Coolthulhu commented:
Adding a damage effect to scorching temperatures (like what is already there for frostbite) would be simple - just 2 lines or so.
Would that be good?
Or is the problem more about heat not accumulating fast enough?

hitbutton commented:
That sounds good.

Rivet-the-Zombie commented:
That's a good, simple solution.

chaosvolt commented:
Should work, yeah.

Coolthulhu commented:
OK, I added the damage effect to scorching heat. Just "hot" isn't enough. Effect is a bit stronger than one from "Freezing" temperature.

However I also found there is an unused mechanic: blisters.
Blisters happen when there are lots of fires nearby.
Unfortunately due to rounding errors, the only way to get blisters in current master is to have 21 "fire levels" in 3x3 block centered on survivor. This means 7 raging fires, 6 raging and 2 medium or 5 raging and 3 medium fires adjacent or under the survivor. That is for a naked survivor - environmental protection 1 would be enough to grant immunity to blisters.
Were blisters supposed to accumulate or something? Like reverse-frostbite?

Some blister mechanics common to "old blisters" and their new implementation:
* Blisters happen regardless of current temperature, only caring for nearby fires and lava.
* Having blisters hurts quite a lot - 1/4 chance of losing a point of hp at blistering part per turn.
* Environmental protection helps fight blisters, as does BARK mutation, heatskin CBM and rm13 armor.
* Blisters have no intensity levels - you either have them or not.
* Blisters affect each body part separately.
* Blisters last roughly 2x as long as the heat that caused them.

Blisters are essentially re-implemented as a part of this PR now.
"New" blisters are calculated as follows:
* Each fire source causes `(power * power) / distance` (rounded down) blisters, where `power` is level of the fire field or 3 in case of lava. Distance of 0 is treated as 1.
* If sum of blisters in a turn is more than `10 + env_resistance on part`, that part is blistered for that one turn
* BARK mutation grants extra 10 blister resistance, heatsink or rm13 grants 20

For example:
* Having 4 medium fires adjacent will cause `(2*2)*4/1=16` blisters - enough to hurt anyone without fire gear, bark or heatsink.
* Stomping around small fires will not cause blisters (due to their low power, they only add 1 blister if adjacent, 0 otherwise).
* 11 tiles of lava, 5 tiles from the survivor is enough to cause blisters to those without environmental gear. Not getting near volcanoes is convectional wisdom.
* A tree-man in entry suit with activated heatsink will need 60 blisters. Standing in the middle of a raging 3x3 pillar of hellfire (or just a big puddle of magma) is `(3*3)*(3*3)/1 = 81` blisters. No swimming in lava.

Also, fixed heat sources increasing (rather than decreasing) frostbite timer in some cases.

chaosvolt commented:
I could've sworn I used to get blisters in older versions, but it seemed to do very little except add to pain, on par with frostnip.

BevapDin commented on src/field.cpp:
```diff
@@ -1836,58 +1830,71 @@ void map::player_in_field( player &u )
+
+                int burn_min = adjusted_intensity;
+                int burn_max = 3 * adjusted_intensity + 3;
                 std::list<int> parts_burned;
```
While you're here, you could change this to be a vector of `body_part` and get rid of the cast at 1884.

BevapDin commented on src/item.cpp:
```diff
@@ -2978,6 +3013,34 @@ int item::chip_resistance( bool worst ) const
+            return stab_resist( to_self );
+        case DT_HEAT:
+            return fire_resist( to_self );
+        default:
```
Why don't you just leave out the default case and let the compiler warn you whenever a new enum value gets added?

Frost-wood commented:
Should add the shaking effect to jittery folks-being a cook I get burned a lot, if you burn a large area I find that body part starts to shake.

Coolthulhu commented:
Thanks for mergetesting.

How is the balancing? It looked OK on my side, but "my side" involved mostly big fires and lava or stomping around fires in fire gear, with some of the typical gear for comparison.
Did I miss anything important?

Rivet-the-Zombie commented:
I like how it turned out; fire is still really dangerous, but it seems to make a lot more sense in terms of *how* it's dangerous. You can fairly readily protect yourself from the immediate effect of flames themselves, but heat will still get you in the end. Everything looks good, @Coolthulhu!
